advisor-backend/routes/schedule.js

```diff
@@ -658,4 +783,245 @@ router.post("/create", async (req, res) => {
   }
 });
 
+// this route handles course swaps into years that are not full of classes already
+router.post("/nonfull", (req, res) => {
+  // schedule array (array of year objects)
+  const schedule = req.body.schedule;
+  // original adjList for the schedule, need this to verify validity of schedule
+  const scheduleAdjList = req.body.scheduleAdjList;
+  // course I want to shift
+  const courseToChange = req.body.courseToChange;
+  // year I want to shift it to
+  const desiredYear = req.body.desiredYear;
+
+  // reverse scheduleAdjList
+  const reverseScheduleAdjList = reverseAdjList(scheduleAdjList);
+
+  // create array of year sets, each set contains the classes for that year
+  const scheduleObject = {};
+  createScheduleObject(schedule, scheduleObject);
+
+  // find year of course to be moved
+  const courseYear = findCourseYear(scheduleObject, courseToChange);
+
+  // nothing to be done
+  if (courseYear === desiredYear) {
+    return res.json({ message: "Swapping into current year" });
+  }
+
+  // if the course's year is before the desired year, the course is being moved forward
+  // for a forward move, need to check if course is coming after it's postreqs
+  if (courseYear < desiredYear) {
+    // find postReqs ////
+    const postReqsSet = new Set();
+    findAllPostrequisites(courseToChange.id, scheduleAdjList, postReqsSet);
+
+    // now to check through the relevant years in scheduleObject for any of the postReqs //
+    const validForwardMove = checkDesiredYearAndBefore(
+      scheduleObject,
+      desiredYear,
+      postReqsSet
+    );
+
+    if (validForwardMove) {
+      // valid move
+      moveCourseToDifferentYear(
+        schedule,
+        desiredYear,
+        courseYear,
+        courseToChange
+      );
+      return res.status(200).json({ message: "Success", schedule });
+    } else {
+      // invalid move
+      return res.status(200).json({ message: "Invalid" });
+    }
+  } else {
+    // if the course's year is after the desired year, the course is being moved back
+    // for a backward move, need to check if course is coming before it's prereqs
+
+    // find preReqs ////
+    const preReqsSet = new Set();
+    findAllPrerequisites(courseToChange.id, reverseScheduleAdjList, preReqsSet);
+
+    // now to check through the relevant years in scheduleObject for any of the postReqs //
+    const validBackMove = checkDesiredYearAndAfter(
+      scheduleObject,
+      desiredYear,
+      preReqsSet
+    );
+
+    if (validBackMove) {
+      // valid move
+      moveCourseToDifferentYear(
+        schedule,
+        desiredYear,
+        courseYear,
+        courseToChange
+      );
+      return res.status(200).json({ message: "Success", schedule });
+    } else {
+      // invalid move
+      return res.status(200).json({ message: "Invalid" });
+    }
+  }
```

Comment:
This also looks really similar and could likely be consolidated into a single chunk of logic.